mirgecom/euler.py

```diff
@@ -20,20 +22,335 @@
+    r"""Computes the inviscid flux vectors from flow solution *q*
+
+    The inviscid fluxes are
+    :math:`(\rho\vec{V},(\rhoE+p)\vec{V},\rho(\vec{V}\otimes\vec{V})+p\mathbf{I})
+    """
+    ndim = discr.dim
+
+    # q = [ rho rhoE rhoV ]
+    qs = split_conserved(ndim, q)
+    mass = qs.mass
+    energy = qs.energy
+    mom = qs.momentum
+
+    p = eos.pressure(q)
+
+    # Fluxes:
+    # [ rhoV (rhoE + p)V (rhoV.x.V + p*I) ]
+    momflux = make_obj_array(
+        [
+            (mom[i] * mom[j] / mass + (p if i == j else 0))
+            for i in range(ndim)
+            for j in range(ndim)
+        ]
+    )
+    massflux = mom * make_obj_array([1.0])
+    energyflux = mom * make_obj_array([(energy + p) / mass])
+    # scalarflux = mom * massfractions / mass
+
+    return flat_obj_array(massflux, energyflux, momflux,)
+
+
+def _get_wavespeed(dim, q, eos=IdealSingleGas()):
+    """Returns the maximum wavespeed in for flow solution *q*"""
+    qs = split_conserved(dim, q)
+    mass = qs.mass
+    mom = qs.momentum
+    actx = mass.array_context
+
+    v = mom * make_obj_array([1.0 / mass])
+
+    sos = eos.sound_speed(q)
+    return actx.np.sqrt(np.dot(v, v)) + sos
+
+
+def _facial_flux(discr, q_tpair, eos=IdealSingleGas()):
+    """Returns the flux across a face given the solution on both sides *q_tpair*"""
+    dim = discr.dim
+
+    qs = split_conserved(dim, q_tpair)
+    mass = qs.mass
+    energy = qs.energy
+    mom = qs.momentum
+    actx = mass.int.array_context
+
+    normal = thaw(actx, discr.normal(q_tpair.dd))
+
+    # Get inviscid fluxes [rhoV (rhoE + p)V (rhoV.x.V + p*I) ]
+    qint = flat_obj_array(mass.int, energy.int, mom.int)
+    qext = flat_obj_array(mass.ext, energy.ext, mom.ext)
+
+    # - Figure out how to manage grudge branch dependencies
+    #    qjump = flat_obj_array(rho.jump, rhoE.jump, rhoV.jump)
+    qjump = qext - qint
+    flux_int = _inviscid_flux(discr, qint, eos)
+    flux_ext = _inviscid_flux(discr, qext, eos)
+
+    # Lax/Friedrichs/Rusanov after JSH/TW Nodal DG Methods, p. 209
+    # DOI: 10.1007/978-0-387-72067-8
+    flux_aver = (flux_int + flux_ext) * 0.5
+
+    # wavespeeds = [ wavespeed_int, wavespeed_ext ]
+    wavespeeds = [_get_wavespeed(dim, qint), _get_wavespeed(dim, qext)]
+
+    lam = actx.np.maximum(wavespeeds[0], wavespeeds[1])
+    lfr = qjump * make_obj_array([0.5 * lam])
+
+    # Surface fluxes should be inviscid flux .dot. normal
+    # rhoV .dot. normal
+    # (rhoE + p)V  .dot. normal
+    # (rhoV.x.V)_1 .dot. normal
+    # (rhoV.x.V)_2 .dot. normal
+    numeqns = number_of_equations(dim, qint)
+    num_flux = flat_obj_array(
+        [
+            np.dot(flux_aver[(i * dim): ((i + 1) * dim)], normal)
+            for i in range(numeqns)
+        ]
+    )
+
+    # add Lax/Friedrichs jump penalty
+    flux_weak = num_flux + lfr
+
+    return discr.project(q_tpair.dd, "all_faces", flux_weak)
+
+
+def inviscid_operator(
+        discr, q, boundaries, t=0.0, eos=IdealSingleGas(),
```

Comment:
Again, do not like defaulted EOS. (I won't repeat this for the functions below, but it applies there, too.)